dot/network/transaction.go

```diff
@@ -119,36 +120,54 @@ func decodeTransactionHandshake(_ []byte) (Handshake, error) {
 	return &transactionHandshake{}, nil
 }
 
-func (s *Service) createBatchMessageHandler(txnBatch chan *BatchMessage) NotificationsMessageBatchHandler {
-	return func(peer peer.ID, msg NotificationsMessage) (msgs []*BatchMessage, err error) {
+func (s *Service) createBatchMessageHandler(txnBatchCh chan *BatchMessage) NotificationsMessageBatchHandler {
+	go func() {
+		protocolID := s.host.protocolID + transactionsID
+		ticker := time.NewTicker(s.cfg.SlotDuration)
+		defer ticker.Stop()
+
+		for {
+		out:
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-ticker.C:
+				timeOut := time.NewTimer(s.cfg.SlotDuration / 3)
+				for {
+					select {
+					case <-timeOut.C:
+						break out
+					case txnMsg := <-txnBatchCh:
+						propagate, err := s.handleTransactionMessage(txnMsg.peer, txnMsg.msg)
+						if err != nil {
+							s.host.closeProtocolStream(protocolID, txnMsg.peer)
+							continue
+						}
+
+						if s.noGossip || !propagate {
+							continue
+						}
+
+						if !s.gossip.hasSeen(txnMsg.msg) {
+							s.broadcastExcluding(s.notificationsProtocols[TransactionMsgType], txnMsg.peer, txnMsg.msg)
+						}
+					}
+				}
+			}
+		}
+	}()
+
+	return func(peer peer.ID, msg NotificationsMessage) {
 		data := &BatchMessage{
 			msg:  msg,
 			peer: peer,
 		}
-		txnBatch <- data
-
-		if len(txnBatch) < s.batchSize {
-			return nil, nil
-		}
 
-		var propagateMsgs []*BatchMessage
-		for txnData := range txnBatch {
-			propagate, err := s.handleTransactionMessage(txnData.peer, txnData.msg)
-			if err != nil {
-				continue
-			}
-			if propagate {
-				propagateMsgs = append(propagateMsgs, &BatchMessage{
-					msg:  txnData.msg,
-					peer: txnData.peer,
-				})
-			}
-			if len(txnBatch) == 0 {
-				break
-			}
+		select {
+		case txnBatchCh <- data:
+		case <-time.After(time.Millisecond * 200):
```

Comment:
Use a timer you stop when the other case triggers instead, this will cause a leak / GC pressure (goroutine won't stop if we get data, and its channel won't be emptied)

Reply:
Done